### ISSUE TITLE
[Feature] Bypass/Disable Warp TAB-completion and use shell TAB-completion

### ISSUE THREAD

**Issue by Soromeister** [warpdotdev/warp#1811]
### Discord username (optional)

Soromeister#1365

### Describe the solution you'd like?

Problem:
When I press TAB I get Warp's dialog pop-up to scroll through the options

Need:
When I press TAB I want to get my shell to do the completion instead of Warp

Reason:
Warp's dialog for TAB completion is just one column and the height is small. 
I want the shell to handle my auto completion by default so when doing a TAB TAB I get the full stuff printed on my STDOUT and it's much more readable this way.

Suggestion:
Make the Warp's TAB completion work with a shortcut, like Option + TAB or ^ + TAB or something similar and under no circumstance pop-up automatically (only when summoned manually through the key sequence).

### Is your feature request related to a problem? Please describe.

I'm frustrated that for mundane things, this Warp TAB completion pops up and won't let my shell complete my command. 
I'd honestly just stick to iTerm2 if it had only the "input as an editor" feature as this is really the only option that is truly valuable in real life. 
All of the other features of Warp just get in the way of doing things and don't make me more productive at all.
Thus, give us the option to completely disable any sort of completion by Warp and let the shell do its job.

### How important is this feature to you?

5 (Can't work without it!)

**Comment by warpdotdev-dx** [comment]
We leverage Fig's specs (we generate the Rust equivalent), and published recent update with 400+ completion specs improvements.


Our completion experience, however, still has lots of room for improvement so thanks for this request. Please +1 this request as it helps us gauge interest.

__
Danny Neira (he/him/his)
Support Engineer
www.warp.dev

**Comment by ozsay** [comment]
@warpdotdev-dx is it (/will it be) possible to add custom specs to Warp?

**Comment by MartinKanters** [comment]
I would love to see a feature like this as well. I think the built-in autocompletion feature is nice, but I don't think it feasible for you to strive for supporting all widely used tools. Next to that a lot of tools people are using might be custom. It would be nice to configure to use either the Warp TAB completion or the shell autocompletion. Even nicer would be to have them both in some way like the author suggested. 
Keep up the good work!

**Comment by elviskahoro** [comment]
The discussion for this is here @MartinKanters @ozsay

**Comment by elviskahoro** [comment]
https://github.com/warpdotdev/Warp/discussions/434

**Comment by jonatan-ivanov** [comment]
>and published recent update with 400+ completion specs improvements.

I don't think the quality of completion support of warp has any relevance here really. I think this is rather about letting the users choose; it does not matter if warp's completion support is good or bad.

**Comment by 0xF4D3C0D3** [comment]
there are a lot of CLI tools and their completions I don't think there could be omniscient tool that covers all these CLI tools

**Comment by dermicus-miclip** [comment]
Why not provide a fallback when no warp/fig completion can be found? I really like the experience so far, but I need AWS CLI daily and this is kind of a buzz kill that I cannot disable it. :(

**Comment by mkubasz** [comment]
@elviskahoro From my perspective, this feature is very crucial. All devops commands currently are very hard to work with them.

**Comment by elviskahoro** [comment]
@mkubasz i think the goal is to make progress on this next quarter. Aloke will probably be spearheading the development of this.

**Comment by mkubasz** [comment]
@elviskahoro Amazing. Sounds great. I really like Warp which is the best term emulator for me!

**Comment by jpbochi** [comment]
Glad to hear there's a plan to do this. I used Warp for a while and really liked it, but felt forced to switch to something else *exclusively* because of this feature.

**Comment by elviskahoro** [comment]
To be clear: there isn't a commitment to supporting shell based completions. The **explorations** as to what we will try to do in order to improve this experience will start probably next quarter--and Aloke will probably take the lead for this.

**Comment by elviskahoro** [comment]
Also the discussion for this is being held here primarily:
https://github.com/warpdotdev/Warp/discussions/652

**Comment by nathanblair** [comment]
I really like Warp and have recommended it but I have to caveat every recommendation with this pain point exists and it doesn't seem like Warp is going in a direction that could make it suitable for my peers in the DevOps space. Between the multitude of benign and popular DevOps tools, to writing our own self-service ones with completions we maintain for `fish`, `zsh`, and `pwsh`, there's no way any clients I work with could adopt Warp and be as productive as they would be in a terminal emulator that doesn't hijack shell autocompletions.

**Comment by alanhg** [comment]
I'm curious, how do tools like Warp and Fig manage to hijack shell autocompletions?

**Comment by ephracis** [comment]
Just to add my two cents here. I have been holding out with Warp for over a year but I finally gave up and went back to iTerm solely due to this missing feature.

Though I miss Warp, completions are such as big part of my day-to-day workflow that I simply cannot be without it.

Please prioritise this. :)

**Comment by jurgob** [comment]
This is the main reason holding me from start to use warp as my main terminal :(

**Comment by P4R** [comment]
> To be clear: there isn't a commitment to supporting shell based completions. The **EXPLORATIONS** as to what we will try to do in order to improve this experience will start probably next quarter--and Aloke will probably take the lead for this.

@elviskahoro Any news about this?

**Comment by tmchow** [comment]
Tab behavior continues to plague me. So much so this is the reason I went back to iTerm2 despite me loving so much of what warp has to offer.  

Given how much ongoing feedback there is on this from the community I am frankly surprised there isn't more progress on this quickly to make this better. Something isn't working.

**Comment by jonatan-ivanov** [comment]
>That being said there is an active effort to improve completions in Warp in general which may include an API for custom completions support.

I've already mentioned this above but I don't think the quality of completion support of Warp has any relevance here really. I think this is rather about letting the users choose; it does not matter if warp's completion support is good or bad.

I think "no it is not possible" or "no we don't want to because X (e.g.: we are planning to commercialize this in the future)" are fair answers but it seems users are leaving because of the lack of this.

**Comment by jtokoph** [comment]
So it seems like the issue here is that Warp takes over all input and doesn't send the input to the underlying shell until you press return. Because of this, there is no way to pass through the tab key to the shell, because the shell wouldn't even know about what was typed at the prompt already.

I feel like the bare minimum would be to allow users to update a blocklist of commands in the Warp config that disable the input hijacking and return to direct shell input mode.

**Comment by synic** [comment]
I think it should just be possible to disable warp tab completion and let the shell do it. I like all the other features, like output bookmarking, copying, AI, but I do not like the completion at all, and it's something that I use often enough that it makes using warp impossible.

Typing `cd ~/Proj` and pressing tab should NEVER suggest anything that doesn't start with `proj`, but the very first suggestion is `Pictures`. In fact, `Projects` is the third option, behind two that don't even match anything other than the `P`. This is not acceptable imho. In the regular shell I can just type `cd ~/Proj`, hit tab, and it's done. It matched the only available option and didn't make me jump through hoops.

**Comment by baroldgene** [comment]
This is quickly becoming a missing feature that is pushing me to abandon Warp.  I really love what has been done but the tab completion is sub-par compared to system tab completion in 2 regards:

- It provides less accurate results: The suggestions are just outright wrong too often to be useful.  As an example If I type `<command> ./` tab completion should only show things in that directory (because ./) and NOT similar commands I've run in entirely different directories that have no bearing here.  
- The keyboard interaction (requiring using arrow keys) is less convenient than the normal "keep pressing tab" style (or whatever has been configured in the shell).  

I think the idea that the shell doesn't know what the user has input is a valid one, but one that can be worked around.  I would think you could query this from the shell pretty easily.  Or create a fake shell, input the command, and see what it would return for tab completion.

**Comment by jpeyret** [comment]
My $.02, as a recent starter with Warp, user of iTerm2 and, mostly, kitty...

This feature isn't for me, as is.

- I write or generate a lot of custom bash/zsh functions to assist with whatever I am working on.  I might for example write a quick launcher `ltestfoo(){ python <path>/testfoo.py $@}` to run whatever failing test I need to fix.  Those are seen just fine by zsh autocomplete, not by warp.  These go into a simple zsh script which I run to initialize each session.

- warp doesn't complete bash variable `foo=1` then `echo $f<tab>` does not work

-  if I `!202` to run 202 from history, `!202` whats now tracked by warp autocomplete.  I want to see what `!202` expanded to.

Bottom line is that some of this needs **local**, user and session-specific, shell-level smarts, not trawling through a database of say kubernetes commands to facilitate gnarly kubernetes work.  Not even my own past sessions, at least not my case, due to my use of throw away bash functions.

What would work is having a shortcut for shell completion and another for warp completion.  Say `tab` goes to warp and `tab tab` in quick succession would go to zsh/bash.  Or vice-versa.

So, while I see the benefit in warp, I'm still spending more time on kitty, when warp features get in the way and cannot be turned off.

**Comment by thedodd** [comment]
When using tools where tab completion will trigger a query to some backing stateful system, such as `kubectl get pod <tab>`, where it will submit a query and present the list of possible pods, or other shell completions with similar functionality — if warp blocks this functionality, then warp is a hard step backwards in capabilities as a product. **Not because warp doesn't have value,** but simply because it is blocking industry standard functionality.

There is still room to innovate, just allow for the standard capabilities of shell completions to be accessed.

**Comment by kensteele** [comment]
> [!NOTE]
> ### I'll preface this by saying I absolutely love what the Warp team has built and the innovative take on reimagining the terminal.

However, aside from open sourcing the client, this has been by far the **_PRIMARY_** issue preventing me from using Warp as my default terminal.  As I'm seeing in a lot of these issues and enhancement requests, a great deal of your users are pleading for standard bash/zsh tab completion due to decades of ingrained muscle memory.  The experience is so jarring that it significantly impacts the speed of command-line navigation which so many of us are accustomed to.

I strongly believe that there is a silent majority of would-be users (**_that would likely quickly become paying customers_**) that simply abandon Warp the moment they attempt to `cd` into a nested path using tab-completion and are presented with the tiny dialog window.  I'd venture to guess that there's likely a second phase of attrition for those that take the time to see if they can disable the feature by toggling `Tab key behavior` -> `Accept autosuggestion` in hopes that it will function like a standard bash/zsh autocomplete.

> [!IMPORTANT]
> ### These users will never see the other benefits Warp has to offer because you've already lost them.

I don't suspect this issue is getting the attention it deserves; it would be very interesting to take a look at the telemetry data and see if you can identify the last feature that was invoked by a new user before abandoning it.  If it's tab-completion or adjusting the `Tab key behavior`, it would be prudent to bump the priority on this or open source enough of the client such that the community can fix it for you.

The top-level  [README](https://github.com/warpdotdev/Warp?tab=readme-ov-file) of this repo states:
> but we are confident that even today the experience is meaningfully better than in other terminals.

This is a bold statement, and I want it to be true.

> [!TIP]
> While taking a deeper look at Warp's plans to open source the client in #400, I came across [waveterm](https://github.com/wavetermdev/waveterm) which is open source and does tab-completion in the way many are requesting right out of the box.

That said, I'll be happy to revisit Warp once issue #1811 is resolved.

**Comment by jpeyret** [comment]
You know, a similar consideration can also be found in modern code editors.  I make extensive use of VS Code snippets *and* I also have CoPilot enabled.  The editor fudges - it is still somewhat kludgy as I write this - a way to let me choose which one to pick.  VS Code *doesn't* say:  "Oh, you use CoPilot?  Well, no more snippets for you".

**Comment by MDGrey33** [comment]
Is there a product manager on this thread? 
If not tag them please.

**Comment by ShepardSiegel** [comment]
I come back to look at Warp every few months for my company to see if this is fixed, it isn't. For all the great Warp features, the brain-damaged, over-aggressive tab-completion is a non-starter. Programmers using Linux all their lives are not going to change for the worse. With all the programming chops Warp has, you would think there could be a "dont be stupid" completion mode that just has to emulate a 1990s shell most of the time. Still hoping for some change, as the rest is nice!

**Comment by dbrxnds** [comment]
I am a new user to Warp and of all good things I see in there, this one immediately makes me quit the application. I think it could work well but the current UX is just to extremely unintuitive that everything I do makes me second guess if that's what I was intending to do.

**Comment by matheuscscp** [comment]
Warp is really great, thank you so much for the amazing innovation here 🙏

But I'd like to stress that it's not about making warp's autocompletion better, but about unblocking an industry standard like someone said above. There's standard tooling out there generating shell scripts for autocompletion for the most popular shells, like bash and zsh. Go's spf13/cobra does that, and that's how countless Go CLI tools including `kubectl` and plenty of other really important DevOps tools generate super smart autocompletion, performing API calls and everything.

Warp simply *must* allow shell autocompletion. I'll come back when this gets implemented.

**Comment by plastikfan** [comment]
Ouch, this issue hurts. I have just discovered the world of completions in particular carapace which I am now looking to integrating in my cli projects. It pains me to discover that my current terminal of choice Warp, kills all of this, so now after using Warp for the past year, looks like will have to find an aletrnative. I can't believe how important completions are, so much so that it rules out the use of Warp as it has declared no intention of allow the shell to perform completions. This is sad, because I love Warp for so many reasons, but this is a show stopper unfortunately. I hope Warp will re-consider.

**Comment by ungarida** [comment]
As many others, I think I can not adopt Warp in my daily workflow because it is really slowing down my workflow. For me is really simple, it is normal and really handy to use shell history. I have a lot of long commands with parameters, a lot of custom scripts, I can easily use every day relying on shell history. With Warp, the experience of browsing the shell history seems broken. I guess I could create a bunch of aliases for each of these commands and that would help to use Warp, but why I should do this extra work for a very basic terminal feature?

**Comment by matheuscscp** [comment]
Any thoughts or updates from the Warp team on this? Is this feature in the pipeline or does it still lack convincing?

**Comment by EmranMR** [comment]
Only thing keeping me away from using warp.

If you are really stuck and want to only use warp, just subshell it, however dont warpify. It does the trick I suppose temporarily... Surely adding this feature (or more correctly, removing) should not be troublesome?

**Comment by tristanlins** [comment]
Our company uses CLI tools with custom autocompletion that works great in Bash and ZSH, but not in Warp. Warp's autocompletion often completes unexpectedly, causing issues. We'd like the option to disable Warp's autocompletion and use our preferred shell's autocompletion instead. Please add a flag to disable Warp's autocompletion.

**Comment by plastikfan** [comment]
> Our company uses CLI tools with custom autocompletion that works great in Bash and ZSH, but not in Warp. Warp's autocompletion often completes unexpectedly, causing issues. We'd like the option to disable Warp's autocompletion and use our preferred shell's autocompletion instead. Please add a flag to disable Warp's autocompletion.

I think you'll be out of luck unfortunately as they have already stated that they will not allow disabling of warp autocomplete in favour of leaving it to the shell. This is a major hindrance to me which is why have abandoned it in favour of Ghostty

**Comment by tristanlins** [comment]
I've moved over to [Tabby](https://tabby.sh/), despite the Electron overhead. Session restoration is a killer feature that many terminals lack, but both Warp and Tabby have it.
It's a shame that user requests often fall on deaf ears. With nearly 150 issues about autocompletion alone, it's clear that this is a significant pain point. I've looked at [this discussion](https://github.com/warpdotdev/Warp/discussions/434), and I get your point.
It seems like Warp is so enamored with its AI features that it's neglecting the basics. Who even uses the AI? Certainly not me.

**Comment by kdocki** [comment]
LLMs in your console seem awesome. I've been doing system administration stuff and working with for over 20 years and after using it for a couple of days I found that I love/hate it. If only the tab would just do the normal bash/shell tab completions, then I would love/love it.

My personal preference would to be leave the tab auto completion up to shell and if I want LLM, then I will type Ctrl+I and ask the questions I need. That would make this tool actually usable for me. 

It seems like such a trivial thing but I can tell you that linux users who have used the terminal for a while will HATE the Tab auto completion override. If there was a setting to turn this off, I'd say take my money every month. 

Keep up the great work guys, when this feature gets added, I'll will sign up for the Team plan.

**Comment by brantburnett** [comment]
I am also trying out Warp now that it's available on Windows, but so far I'm finding it very difficult to use. We have a lot of PowerShell scripts that include complex parameters with tab completion support via `ValidateSet` for the values. However, in PowerShell I can't use the autocompletion for the parameter names or parameter values from these scripts when using Warp. Completion libraries will never get support for this kind of thing, as it depends on parsing the target PowerShell script file. I love the other features of Warp, but this is such a large friction point for me that, so far, it's outweighing the benefits of the other features.

**Comment by RyotaMono** [comment]
I believe the wrap team doesn't care about providing a bypass mode at all. There are tens, if not hundreds, of posts requesting this feature, but there have been no meaningful responses for nearly THREE YEARS. They even keep it as part of the AI subscription. This is the kind of terrible function that disrupts users' daily workflows and immediately drives them away.

**Comment by dannyneira** [comment]
Hey folks, sorry for the silence. This feature, of supporting shell-based completions/plugins/functions while maintaining Warp's features like Blocks, IDE-like input, Agent Mode, etc., is quite a technical challenge. Given that there are nearly an endless number of shell plugins and many shells to support, I hope you can understand the intricacies involved.

That being said, we do listen to feedback and given enough interest in the feature, we have [completed many of the top requests our users have submitted over the years](https://github.com/warpdotdev/Warp/issues?q=is%3Aissue%20state%3Aclosed%20%20sort%3Areactions-%2B1-desc) (even challenging ones like [Warp on Windows](https://www.warp.dev/blog/building-warp-on-windows)).

To anyone else interested in this feature, please add a 👍 to the [original post at the top](https://github.com/warpdotdev/Warp/issues/1811#issue-1364739609) to signal that you want this feature, and subscribe if you'd like to be notified.

**Comment by olinox14** [comment]
@dannyneira IMO the ideal case would be to be able to get the native shell auto-completion with tab, and the warp AI powered one with another control, like shift+tab or whatever other shortcut. But this is indeed the only and one thing I don't like about warp today, hope you'll find a solution. Thanks for the feedback.

**Comment by AlphaComposite** [comment]
This was a very silly and arbitrary decision to impose on your users. It's one of the dumbest things a product has ever done. This is a command utility, and you're taking away the TAB key when many of us DEPEND on that for our own completions. 

FIX THIS.

**Comment by mscurtescu** [comment]
Can this be added to known issues and/or FAQ?

Also, switching to agent mode and asking about tab completion you are reassured that it's working, very confusing.

**Comment by wfy-belief** [comment]
+1 The experience was very bad when I used telepresence who is a k8s tool

**Comment by sdhull** [comment]
@dannyneira curious what you think about @olinox14's comment, specifically:
> IMO the ideal case would be to be able to get the native shell auto-completion with tab, and the warp AI powered one with another control, like shift+tab or whatever other shortcut

Why not just use some other key binding for the Warp functionality currently attached to `tab`? Why is this such an enormous technical challenge? Feel free to take over `tab` while in Agent Mode (since Agent Mode is not a terminal). Not sure why Blocks make this more difficult and not even sure what you _mean_ by "IDE-like input."

> we have [completed many of the top requests our users have submitted over the years](https://github.com/warpdotdev/Warp/issues?q=is%3Aissue%20state%3Aclosed%20%20sort%3Areactions-%2B1-desc)
Just going by 👍's, this would be #10 on that list, and tied for #4 on the list of open issues. Can you give us some idea of whether or not this is being worked on or at least prioritized for the near future?

**Comment by dannyneira** [comment]
It's a challenge technically since we'd have to support a number of shell-based completions, bindkey, etc., and have it all work with Warps features.

When I mentioned IDE-like input, it's because you can use the mouse to move the cursor around, it has syntax highlighting by default, you can click-drag and highlight/cut/copy text, you can have multiple cursors, etc., many things that most terminal inputs don't support but IDEs do.

That being said, the more interest in the feature, the higher chance of priority in implementation as evidenced by our [closed issues](https://github.com/warpdotdev/Warp/issues?q=is%3Aissue%20state%3Aclosed%20sort%3Areactions-%2B1-desc), so please continue to 👍  the [issue](https://github.com/warpdotdev/Warp/issues/1811#issue-1364739609) to show interest.

Thanks for your patience as we work to improve Warp.

**Comment by AlexMaskovyak** [comment]
This is a big enough feature gap that I'm abandoning Warp. I loved it for a bit, but it's janky auto-completions are actively causing problems.

**Comment by ElijahLynn** [comment]
In the meantime, could Warp at least give some user feedback when I am mashing tab and nothing is happening? e.g. "this doesn't work the way you think it should, see ".

**Comment by mtwomey** [comment]
+1 on this. Just realized the systemctl completions don't work in Warp and it's frustrating to have to stop and lookup the correct names of the services I'm trying to deal with.

**Comment by ttnghia** [comment]
I have zsh auto suggestion works really well. If Warp can't suggest, please provide an option to disable the suggestion feature so I can switch to my zsh one.

**Comment by christian-taillon** [comment]
+1 on this request. This is needed. 

I'd like to add my support for this feature request as a Warp Pro customer. I purchased Pro but honestly find it so disruptive to my work flows. 

**Request:**  
Please provide an option to fully disable the popup dialog that shows folders and files—whether triggered by `@` or when navigating paths. While I understand the intent behind these ergonomic features, for my workflow, these popups are a significant distraction and actually reduce my productivity. I prefer to use my shell's native navigation and completion, and would rather not have any UI overlays for file/folder selection, even if it means losing some of the enhanced dialog options.

**Why this matters:**  
- The popup appears even when AI and related features are disabled, and there is currently no setting to turn it off ([source](https://github.com/warpdotdev/Warp/issues/6618), [source](https://github.com/warpdotdev/Warp/issues/6656)).
- This behavior interferes with common development workflows, such as using `@` in version control (e.g., Jujutsu) or package management (e.g., npm, yarn workspaces).
- Hitting `ESC` to close the popup each time is not a viable long-term solution.

**Suggestion:**  
Please consider making all file/folder popups (not just those triggered by `@`) fully optional, ideally with a simple toggle in the settings. I am happy to lose some dialog features if it means a cleaner, distraction-free terminal experience.

Thank you for considering this!

**Comment by omer-ibex** [comment]
please resolve this matter ASAP as users are not using WARP because of it.
it is a natural and simple thing to keep the bash auto-complete feature and not "run over" it with another feature.
when i "cd", id like to begin typing the folder i want to reach and hit "tab" to receive an auto-complete from terminal.
At the moment this is not possible with WARP so I am abandoning the product for now.

**Comment by dannyneira** [comment]
hey folks,
We understand the interest in this feature, and we'll post updates on this thread if/when it's on the roadmap. Note that it will require significant refactoring and redesign of Warp's wrapper in order to get this to work. Since Warp has a wrapper around the shell to enable features like Blocks, Completions, IDE-like input editor, AI integration, etc., so in order to support shell based completions and plugins we're need to find a way to bypass that wrapper.

For now the best workaround is to run your shell in a subshell (you'll lose the Warp features, but your shell completions should work as normal).

Please continue to 👍  the [top line request](https://github.com/warpdotdev/Warp/issues/1811#issue-1364739609) to continue to show interest (unfortunately we don't count comment +1's).

**Comment by dannyneira** [comment]
For example run 'zsh', then don't warpify the session and use the zsh shell as a subshell.

**Comment by dannyneira** [comment]
@ledniy that's essentially what this request is about, maintain the Warp experience, but with shell completions. And that is what would require significant engineering and design work to accomplish for the reasons I mentioned above. We appreciate your patience with this.

**Comment by JoshuaHintze** [comment]
I'm finding that a few completions I normally expect to work do not. For example, I use oh my zsh plugin (pj) project jump. In normal terminal sessions I can tab complete to see these project folders. 

I also have noticed that you can't tab complete symbol link to folders either.

**Comment by lucieleblanc** [comment]
Hi folks, we've been closing specific completion gaps raised in this thread. These don't address the core request (native shell completion passthrough), but cover many of the tools people called out:

* **AWS CLI**: improved top-level completions with profile/region generators + added dynamic completions for EC2 and S3 (@dermicus-miclip)
* **kubectl**: fixed `--context` and `--cluster` persistent flag completions + added `$KUBECONFIG` env var support (@thedodd, @DerrickMartinez, @wfy-belief)
* **Homebrew**: added formulae and cask name completion for `brew install`, `brew uninstall`, etc.
* **Docker**: added `docker-compose` service name completion + `docker run` image:tag completion
* **gcloud**: added dynamic generators for zones, regions, components, and `gsutil` bucket/object completions (@kvokka)
* **DevOps tools**: improved advanced completions for Terraform + added support for Azure CLI, OpenShift (@mkubasz, @nathanblair, @kensteele)

All of the above are available either in last week's release (`v0.2026.04.08.08.36`) or will be in this week's release in the next few days. Thank you for all the requests, and please continue to open individual issues for tools you'd like to see supported.

**Comment by wspear** [comment]
Interesting approach... is there (or could there be) a way to feed an application's auto-complete spec to warp and have it take over that functionality, bypassing native completion but not relying on new releases to support individual tools?

**Comment by plastikfan** [comment]
As much as I appreciate this effort, it doesn't help the completion for non main stream tools. I'm working on building some cli tools that have zero traction which means you can't support it. The sooner we just get shell completion fully implemented, the better.

**Comment by matheuscscp** [comment]
I don't even use CLIs myself anymore, this issue can be closed lol

**Comment by kvokka** [comment]
@lucieleblanc, as it was mentioned earlier, it will not cover everything. And a new tool will appear, where the developer might have autocomplete in other terminals except Warp. The solution that you are giving is yet another spike.

It's cool that most of the users are ok with the default basic autocomplete that Warp provides, but it must be extensible. Let the user somehow set the autocompletes that he/she wants, but now it's like all or nothing. And for those who are not ok with the default preset, the best option is to pick another terminal.

**Comment by jayralbano** [comment]
its frustrating. i'm expecting file name completion, and this is what i get lol

<img width="399" height="94" alt="Image" src="https://github.com/user-attachments/assets/6d30b71a-e0ab-4da9-81ac-c51b737af213" />

**Comment by JoshuaHintze** [comment]
I stopped using warp and went back to Windows Terminal and I forgot how much I missed the auto completions actually working. I also noticed with zsh-autocompletion plugin that looks at your history and allows you to press right arrow that its almost the same as the AI completion warp has so I feel like I'm missing nothing. When I want to ask AI I'll just ask claude code to do somethign in my terminal.

**Comment by ElijahLynn** [comment]
Now that Warp is open source, https://www.warp.dev/blog/warp-is-now-open-source, I have confidence the open source community can deliver a workaround here.

**Comment by mboeker** [comment]
My workaround for this has been to just run `bash` first on every warp terminal. I'm still trying to find a way to automate this. So far the things I've but in my dotfiles have had funky side-effects..

**Comment by tristanlins** [comment]
## Native shell completions already exist and work — zsh only, behind a hidden flag

Since the code is public now: Warp already contains a full implementation that hands completion off to your real shell. It writes `^Y` to the PTY, a widget installed by Warp's own zsh bootstrap runs the genuine `compsys` pipeline (`_generic`), and an overridden `compadd` streams the matches back over an OSC sequence — so your `compinit`, your plugins and your custom `_my_tool` functions all apply. Results are then rendered in Warp's popup.

I enabled it on stable `v0.2026.07.29.09.05.stable_02` (Linux, zsh) and it works: completions start coming from my own zsh config, including for tools Warp has no spec for.

Two behaviours exist in the code: the feature flag alone treats the shell as a *fallback* (Warp's specs win, the shell fills gaps), while the private `ForceNativeShellCompletions` preference makes the shell win outright. The latter is what most of this thread is asking for.

**How to enable it** (quit Warp first — the store is read once at startup and overwritten on the next write):

- **Linux** — merge into `~/.config/warp-terminal/user_preferences.json`: `{"prefs":{"ForceNativeShellCompletions":"true"}}`. The `prefs` wrapper is required and the value must be the **string** `"true"`; either mistake fails silently and discards *all* private preferences.
- **macOS** — `defaults write dev.warp.Warp ForceNativeShellCompletions -string true`
- **Windows** — `HKEY_CURRENT_USER\Software\Warp.dev\Warp`, `REG_SZ` named `ForceNativeShellCompletions` = `true`

Limits: zsh only, skipped for multi-line commands, and the candidates render in Warp's menu rather than as shell output. The `TAB TAB → columns on STDOUT` part of the original request stays impossible, and not for lack of effort: the prompt is a GUI editor, the shell receives nothing until Enter, and then the line arrives as one bracketed-paste burst preceded by a kill-line. There is no partial line for the shell to complete. That is what makes blocks work.

### Related

- #4521 — same underlying problem: fall back to shell-native completions when Warp has no spec
- #13778 — asks for exactly this opt-in
- #2694 — the private preference is mentioned in passing

### This should not have taken five years and a source-code release to find out

What I don't understand is the communication. In #13778, Warp's own triage bot wrote on 2026-07-15:

> Native shell completion plumbing already exists, but it is gated by `FeatureFlag::NativeShellCompletions` or the private `ForceNativeShellCompletions` preference and is currently used as a fallback/supplement rather than as a public setting.

That is an accurate description of a working feature, posted in a mac-specific issue with `needs-info` and zero reactions. Meanwhile *this* thread — 370 👍, five years old, the canonical place where people ask for this — got a list of individual completion-spec fixes in April with the note that they "don't address the core request (native shell completion passthrough)". Nobody mentioned that a working zsh implementation had been sitting in the build since at least the open-source release, one preference key away.

Several people in this thread have cancelled subscriptions or left for other terminals over precisely the gap this feature closes. A single sentence here at any point in the last years would have changed that for the zsh users among them.

So: please make it a real setting, and document it. Given the code already distinguishes "supplement" from "prefer the shell", it wants to be a three-way choice rather than a checkbox. But the more urgent ask is simply to say out loud what already exists.

**Comment by mboeker** [comment]
> ## Native shell completions already exist and work — zsh only, behind a hidden flag
> Since the code is public now: Warp already contains a full implementation that hands completion off to your real shell. It writes `^Y` to the PTY, a widget installed by Warp's own zsh bootstrap runs the genuine `compsys` pipeline (`_generic`), and an overridden `compadd` streams the matches back over an OSC sequence — so your `compinit`, your plugins and your custom `_my_tool` functions all apply. Results are then rendered in Warp's popup.
> ... 
> * **Windows** — `HKEY_CURRENT_USER\Software\Warp.dev\Warp`, `REG_SZ` named `ForceNativeShellCompletions` = `true`

This does not work for me at all...
I added the registry key (windows). Latest Warp: launch an Ubuntu shell. Type "lsb_r" and press tab, nothing happens. In faded font it shows "lsb_release" but it does not complete it.

**Comment by ryofurue** [comment]
My issue is [Tab key inserts spaces instead of performing bash completion #6562](https://github.com/warpdotdev/warp/issues/6562) but it's marked as duplicate of this issue.

To me, Warp's own completion/suggestion is fine, but when Warp doesn't know the command
```
$ commandname<SPACE><TAB>
```
inserts spaces instead of trying to complete for filename, which is the biggest annoyance to me.  That has made me try to bypass Warp's completion (and finding it impossible, I have temporarily abandoned Warp), but that's not my goal. I just want the TAB key to always invoke the completion function.

In this sense, I don't think it's a good idea to mark #6562 as duplicate of this thread.

Or has this thread been "converted" to a catch-all thread about various issues of completion?

**Comment by acarl005** [comment]
I'm currently working on this. Is just experimental at the moment so no guarantees on anything releasing anytime soon

**Comment by acarl005** [comment]
We're launch support for this tomorrow in the [Warp Preview build](https://www.warp.dev/download-preview). By default Warp will run its own existing completions. If it doesn't have anything, it will fallback to asking the shell. You can go to Settings > Features > Terminal Input and there will be 2 settings, "Warp completions" and "native shell completions", both enabled by default. Disable "Warp completions" and then you will only get completions from the shell. Please try it out tomorrow after the update.

UPDATE: This is released on Preview version `v0.2026.09.02.08.27.preview_01`

**Comment by AlexMaskovyak** [comment]
Shouldn't it perform shell completion first and then fallback to Warp
completions? I mean, auto-completing a path on the filesystem should be
deterministic and based upon what's authoritative for your local shell.

Best,
-Alex Maskovyak

**Comment by acarl005** [comment]
@AlexMaskovyak Good question. Due to Warp's architecture, native shell completions have to be more complicated than a typical terminal, and there is serialization overhead. For file/folder completions specifically it's faster to do it in Warp's own system.

**Comment by jonatan-ivanov** [comment]
@acarl005 Did you mean to close this issue or was it an incident?

I would recommend keeping it open to gauge user interest since this issue with its duplicates seem to be the second highest voted one at the moment. If you really want to close it, I would recommend closing it with the "Close as not planned" button that reflects the intention: the issue is not fixed and there are no plans fixing it.

**Comment by amannan-123** [comment]
@jonatan-ivanov I think he fixed it in the alpha version. Check the above comments.
